### PR TITLE
Add tpm2tools

### PR DESCRIPTION
**Pull request by josephlr**
This package is designed to provide helper methods to make dealing with primary keys easier.

We also update our tests to use these helper functions.

### CODE REVIEW

File: tpm2tools/keys_test.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	}
+
+	hashFn, err := ek.pubArea.NameAlg.HashConstructor()
+	if err != nil {
+		t.Fatal(err)
+	}
+	hash := hashFn()
+
+	hash.Write(pubEncoded)
+	if !bytes.Equal(hash.Sum(nil), ek.Name().Value) {
```
Comment (twitchy-jsonp) [inline tpm2tools/keys_test.go]: use `tpm2.Name.MatchesPublic`?

File: tpm2tools/keys.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+// when finished using the Key, so that the underlying TPM handle can be freed.
+type Key struct {
+	rw           io.ReadWriter
+	handle       tpmutil.Handle
+	pubArea      tpm2.Public
+	pubKey       crypto.PublicKey
+	creationData *tpm2.CreationData
+	creationHash []byte
+	ticket       *tpm2.Ticket
+	name         *tpm2.HashValue
```
[the remark this answers is not in the document]
Comment (josephlr) [reply tpm2tools/keys.go]: Done

File: tpm2tools/keys.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+// EndorsementKeyRSA returns the key created from DefaultEKTemplateRSA.
+func EndorsementKeyRSA(rw io.ReadWriter) (*Key, error) {
+	return EndorsementKeyFromTemplate(rw, DefaultEKTemplateRSA())
+}
+
+// EndorsementKeyFromNvIndex loads an endorsement key using the template
```
[the remark this answers is not in the document]
Comment (josephlr) [reply tpm2tools/keys.go]: So CreatePrimary both generates a key and loads it. Comments changed to make this clearer

File: tpm2tools/keys.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+		return nil, fmt.Errorf("read error at index %d: %v", idx, err)
+	}
+	template, err := tpm2.DecodePublic(data)
+	if err != nil {
+		return nil, fmt.Errorf("index %d data was not a TPM key template: %v", idx, err)
+	}
+	return EndorsementKeyFromTemplate(rw, template)
+}
+
+// EndorsementKeyFromTemplate loads a primary key in the endorsement hierarchy
```
[the remark this answers is not in the document]
Comment (josephlr) [reply tpm2tools/keys.go]: Good idea. API changed to have a NewKey method which takes in a parent. Later we can add support for non-primary keys without introducing a new API.

File: tpm2tools/keys.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+// using the provided template. This function assumes that the desired key:
+//   - Does not have its usage locked to specific PCR values
+//   - Usable with empty authorization sessions (i.e. doesn't need a password)
+func StorageKeyFromTemplate(rw io.ReadWriter, template tpm2.Public) (*Key, error) {
+	return newPrimaryKey(rw, tpm2.HandleOwner, template)
+}
+
+// Wrapper around CreatePrimary used to load a key which does not need to be
+// locked to any PCRs or use any authorization sessions.
+func newPrimaryKey(rw io.ReadWriter, owner tpmutil.Handle, template tpm2.Public) (*Key, error) {
```
[the remark this answers is not in the document]
Comment (josephlr) [reply tpm2tools/keys.go]: Fixed to use defer, Close(), and implicit return

File: tpm2tools/keys_test.go (hunk 1 above)
Comment (josephlr) [reply tpm2tools/keys_test.go]: Done

**Comment by josephlr** [comment]
@twitchy-jsonp this should now be good for a full review. Tests and docs have been added.

File: tpm2tools/template.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		NameAlg:    tpm2.AlgSHA256,
+		Attributes: tpm2.FlagSignerDefault,
+		RSAParameters: &tpm2.RSAParams{
+			Sign: &tpm2.SigScheme{
+				Alg:  tpm2.AlgRSASSA,
+				Hash: tpm2.AlgSHA256,
+			},
+			KeyBits:    2048,
+			Exponent:   0,
+			ModulusRaw: nonce[:],
```
Comment (twitchyliquid64) [inline tpm2tools/template.go]: Whats happening here? is the the public.unique value?

Just want to double check that `ModulusRaw` is not actually going to end up being the modulus (as that has security-relevant properties for the generated key).

File: tpm2tools/keys_test.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	"testing"
+
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+
+	"github.com/google/go-tpm-tools/simulator"
+)
+
+func getEK(t *testing.T) (*simulator.Simulator, *Key) {
+	simulator, err := simulator.Get()
```
Comment (twitchyliquid64) [inline tpm2tools/keys_test.go]: `t.Helper()`

Comment (josephlr) [reply tpm2tools/keys_test.go]: Done

File: simulator/simulator_test.go [hunk 8]
```diff
@@ -30,19 +34,43 @@ func getSimulator(t *testing.T) *Simulator {
 	return simulator
 }
 
-func TestReset(t *testing.T) {
+func getEKModulus(t *testing.T, rwc io.ReadWriteCloser) *big.Int {
+	ek, err := tpm2tools.EndorsementKeyRSA(rwc)
```
[the remark this answers is not in the document]
Comment (josephlr) [reply simulator/simulator_test.go]: Done, here and above

**Pull request merged**